### DISCUSSION TITLE
Strange behavior when moving elements in an array.

### DISCUSSION THREAD

**Discussion by Dzemy** [livewire/livewire#7402] [1. Help]
Hey, everybody.
I need help with a very simple code, I'm solving it for the second day and I've reached a point where I've simplified the code a lot and I still don't understand its behavior.

Here is a simplified version of my application.

```php
class Test extends Component
{
    public $array = [
        [
            'name' => 'Alpha',
            'id' => '1001',
            'email' => 'alpha@example.com'],
        [
            'name' => 'Bravo',
            'id' => '1002',
            'email' => 'bravo@example.com'],
         ...
    ];

    public function updating()
    {
        //dd('updating');
    }

    public function updated()
    {
        //dd('updated');
    }

    public function addTop()
    {
        array_unshift($this->array, [
            'name' => 'Top',
            'id' => '9999',
            'email' => 'test']);
    }

    public function moveDown($key)
    {
        [
            $this->array[$key],
            $this->array[$key + 1]] = [
                $this->array[$key + 1],
                $this->array[$key]];
    }

    public function render()
    {
        return view('testing::livewire.test');
    }
}
```

```blade
<div>
    <button class="btn" wire:click="addTop">Add</button>
    @foreach($array as $key => $arr)
        <div wire:key="{{ $arr["id"] }}">
            <input class="input" wire:model.live="array.{{$key}}.name">
            <button class="btn" wire:click="moveDown({{$key}})">Down</button>
        </div>
    @endforeach
</div>
```

Now where is the problem, well firstly when I press the "Add" or "Down" button the updated|updating function is not called. The updated function is only called when I overwrite a value in the input, thanks to wire:model. What is interesting then is that once I call updated using the input, the buttons start calling this function too until I refresh the page.

This is the first problem, the second one is related to wire:key, if I use it as it is in the code, when I move the elements, the bindings between the models start to appear.

![image](https://github.com/livewire/livewire/assets/21079305/39549694-3ddc-4b61-a164-948451974cdb)

If I remove the wire:key, the bindings are not created, but according to the documentation it is necessary to insert a unique key into the wire:key when using loop.

```
PHP = 8.2
Livewire = 3.2.1
Laravel = 10.33.0
```

**Comment by Progression** [comment]
@Dzemy Did you find a solution to this. I'm seeing the same issue after upgrading to Livewire V3. 

Any sorting of an array (or collection) of arrays or objects, which are bound to input controls results in the input controls all mirroring each other. I too have simplied this down to the most basic case in a new project and it happens everytime and I've tried sorting or rebuilding the underlaying array data in various ways and nothing works.

**Reply by joshhanley** [reply]
@Progression any chance you could submit a failing test PR that demonstrates this issue? Thanks!

**Reply by Progression** [reply]
I can try! Never submitted a PR before but I'll try and give it ago over the weekend.

**Reply by joshhanley** [reply]
@Progression thanks! 🙏 if you have any troubles, let me know 🙂 We have a contribution guide here which should help https://livewire.laravel.com/docs/contribution-guide

**Reply by Progression** [reply]
Followed the contribution guides but when I run the tests it has a lot of errors. BOth dusk related ones later on (I've closed Chrome related apps - should this include things like Electron apps?) but earlier issues around SQL...

```
1) Livewire\Features\SupportFileUploads\UnitTest::the_global_upload_route_middleware_is_configurable
Failed asserting that 'SQLSTATE[HY000]: General error: 1 no such table: sessions (Connection: testbench, SQL: select * from "sessions" where "id" = aYFbLVuO2iXQ5FIAyJXV4BFP8AUJGfogAG0XgjNN limit 1)' [ASCII](length: 174) contains "Livewire\Features\SupportFileUploads\DummyMiddleware" [ASCII](length: 52).

/Users/robclaisse/Projects/packages/livewire/src/Features/SupportFileUploads/UnitTest.php:436
```
Any help to speed up getting this setup so I can get on to actually creating the test to show the input binding bug?

**Reply by joshhanley** [reply]
@Progression there was a bug with Laravel 11 installs, this has been fixed in PR #7874, so if you pull the latest changes from `main` it should be fixed 🙂

**Reply by Progression** [reply]
FYI for anyone reading this I have now submitted a PR with a failing test to illustrate the issues: #7888
